bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTable.java

```diff
@@ -0,0 +1,1411 @@
+    inOrder.verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+
+    verify(primaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+    verify(secondaryTable).batch(eq(Arrays.asList(put1)), any(Object[].class));
+  }
+
+  @Test
+  public void testNonConcurrentOpsWontBePerformedConcurrently()
+      throws IOException, InterruptedException {
+    setupMirroringTableWithDirectExecutor();
+    Get get = createGet("get1");
+    Increment increment = new Increment("row".getBytes());
+    Append append = new Append("row".getBytes());
+
+    Put put = createPut("test1", "f1", "q1", "v1");
+    Delete delete = createDelete("test2");
+
+    mockBatch(
+        primaryTable,
+        secondaryTable,
+        get,
+        createResult(),
+        increment,
+        createResult("row", "v1"),
+        append,
+        createResult("row", "v2"));
+
+    checkBatchCalledSequentially(Arrays.asList(get));
+    checkBatchCalledSequentially(Arrays.asList(increment));
+    checkBatchCalledSequentially(Arrays.asList(append));
+
+    checkBatchCalledConcurrently(Arrays.asList(put));
+    checkBatchCalledConcurrently(Arrays.asList(delete));
+    checkBatchCalledConcurrently(Arrays.asList(put, delete));
+
+    checkBatchCalledSequentially(Arrays.asList(put, delete, get));
+    checkBatchCalledSequentially(Arrays.asList(put, delete, increment));
+    checkBatchCalledSequentially(Arrays.asList(put, delete, append));
+  }
+
+  private void setupMirroringTableWithDirectExecutor() {
+    this.mirroringTable =
+        spy(
+            new MirroringTable(
+                primaryTable,
+                secondaryTable,
+                MoreExecutors.listeningDecorator(MoreExecutors.newDirectExecutorService()),
+                mismatchDetector,
+                flowController,
+                secondaryWriteErrorConsumer,
+                new ReadSampler(100),
+                true,
+                new MirroringTracer()));
+  }
+
+  private void checkBatchCalledSequentially(List<? extends Row> requests)
+      throws IOException, InterruptedException {
+    InOrder inOrder = Mockito.inOrder(primaryTable, flowController, secondaryTable);
+    this.mirroringTable.batch(requests, new Object[requests.size()]);
+    inOrder.verify(primaryTable).batch(eq(requests), any(Object[].class));
+    inOrder.verify(flowController).asyncRequestResource(any(RequestResourcesDescription.class));
+    inOrder.verify(secondaryTable).batch(ArgumentMatchers.<Row>anyList(), any(Object[].class));
+  }
+
+  private void checkBatchCalledConcurrently(List<? extends Row> requests)
+      throws IOException, InterruptedException {
+    InOrder inOrder = Mockito.inOrder(primaryTable, flowController, secondaryTable);
+    this.mirroringTable.batch(requests, new Object[requests.size()]);
+    inOrder.verify(flowController).asyncRequestResource(any(RequestResourcesDescription.class));
+    inOrder.verify(secondaryTable).batch(eq(requests), any(Object[].class));
```

Comment:
Is it guaranteed to hit secondary first then primary? Is this something to do with DirectExcecutor? A comment explaining the algo, will be helpful.

Reply:
Done.